src/main/java/hudson/remoting/Channel.java

```diff
@@ -159,7 +159,8 @@
      * Requests that are sent to the remote side for execution, yet we are waiting locally until
      * we hear back their responses.
      */
-    /*package*/ final Map<Integer,Request<? extends Serializable,? extends Throwable>> pendingCalls = new Hashtable<>();
+    /*package*/ final Map<Integer,Request<? extends Serializable,? extends Throwable>> pendingCalls =
+            Collections.synchronizedMap(new HashMap<>());
```

Comment:
Any reason not to use ConcurrentHashMap here?

Actually, if you look at the usages of pendingCalls, I'm pretty certain it's thread safety is broken, before and after. Especially if we accept that the synchronized block near 1083 is important.

Reply:
Code that called this was synchronizing on the collection itself which would require more refactoring to support properly. The synchronized views all use the collection as the lock which works as a drop in replacement for Vector and Hashtable for the most part.

Reply:
Sometimes the calling code is synchronizing on the collection, but not always. Here you changed it from a Hashtable to a synchronizedMap. I'm not sure that helps and it might hurt.

I'm concerned about introducing more synchronization in an area where it isn't clear.

Reply:
To be more clear: you only need to synchronize on the collection to do atomic operations. You don't need to synchronize on it for get and put for example, but if you wanted to iterate over the entries, you should synchronize then to avoid concurrent modifications.